common/selectors/selectAndDedupe.js

```diff
@@ -13,11 +10,15 @@ dedupe.defaults.createKey = site => (site ? site.cache_key || site.hostname || s
  * @param {Array} group.defaults - default values to use.
  * @returns {Array}
  */
-module.exports = function selectAndDedupe(group) {
+module.exports.selectAndDedupe = function selectAndDedupe(group) {
   return group.reduce((result, options, index, arr) => {
     let current;
     let sites = options.sites;
 
+    // NOTE: The Math.random() fallback is because there can be empty slots thanks to pinning
+    // with a limited history.
+    dedupe.defaults.createKey = site => (site ? site.cache_key || site.hostname || site.url : Math.random());
```

Comment:
Yeah, this is a bug right now. deduping needs to move into the feed for top sites and happen before the pinned sites get all sorted, etc. But that might not happen until system addon porting.